nominal/nominal.py

```diff
@@ -1,628 +1,316 @@
-        }
-
-    def __upload_file(self, overwrite: bool) -> requests.Response:
-        """
-        Uploads dataframe to S3 as a file.
-
-        Returns:
-        Response object from the REST call.
-        """
-
-        if self.s3_path is not None and not overwrite:
-            print(
-                "\nThis Dataset is already uploaded to an S3 bucket:\n{0}\nTry [code]upload(overwrite = True)[/code] to overwrite it.".format(
-                    self.s3_path
-                )
-            )
-            return
-
-        # Create a default dataset name
-        if self.filename is None:
-            self.filename = default_filename("DATASET")
-
-        csv_file_buffer = io.BytesIO()
-        self.write_csv(csv_file_buffer, datetime_format=r"%Y-%m-%dT%H:%M:%S.%fZ")
-
-        # Get the size of the buffer in bytes
-        csv_file_buffer.seek(0, os.SEEK_END)
-        csv_buffer_size_bytes = csv_file_buffer.tell()
-        csv_file_buffer.seek(0)
-
-        print(
-            "\nUploading: [bold green]{0}[/bold green]\nto {1}\n = {2} bytes".format(
-                self.filename, get_base_url(), csv_buffer_size_bytes
-            )
-        )
 
-        # Make POST request to upload data file to S3
-        resp = requests.post(
-            url=ENDPOINTS["file_upload"].format(get_base_url(), self.filename),
-            data=csv_file_buffer.read(),
-            params={"sizeBytes": csv_buffer_size_bytes},
-            headers=self.__get_headers(content_type="octet-stream"),
+def get_default_connection() -> NominalClient:
+    """Retrieve the default global connection to the Nominal platform.
+
+    Raises nominal.exceptions.NominalError if no global connection has been set.
+    """
+    global _default_connection
+    if _default_connection is None:
+        raise NominalError("No default connection set: initialize with `set_default_connection(base_url, token)`")
+    return _default_connection
+
+
+def upload_dataset_from_pandas(
+    df: pd.DataFrame,
+    name: str,
+    description: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a pandas.DataFrame."""
+    conn = get_default_connection()
+
+    # TODO(alkasm): use parquet instead of CSV as an intermediary
+
+    def write_and_close(df: pd.DataFrame, w: BinaryIO) -> None:
+        df.to_csv(w)
+        w.close()
+
+    with reader_writer() as (reader, writer):
+        # write the dataframe to CSV in another thread
+        t = Thread(target=write_and_close, args=(df, writer))
+        t.start()
+        dataset = conn.create_dataset_from_io(
+            reader,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=FileTypes.CSV,
+            description=description,
+        )
+        t.join()
```

Comment:
to confirm the `reader_writer` is here to proxy logs from the other thread to this main one?

also, wondering if we need the complexity of a separate thread given this method blocks on the upload and ingest anyway, so the main thread will always be waiting for the thread to complete

Reply:
the multithreading isn't for speed, but rather to avoid copying the dataframe to an in-memory string